Add error propagation to interpreter creation logic

anutosh491 opened the pull request:
# Description

@Vipul-Cariappa and I were discussing a use case yesterday (especially while building with repl) and this is what we realized

1) The function responsible for creating `clang::Interpreter` is `createClangInterpreter` which can return a nullptr, for eg 
https://github.com/compiler-research/CppInterOp/blob/c3384fb9d784a3f1606b6fc20c283c1e48d9b14b/lib/Interpreter/Compatibility.h#L277

2) This is assigned to `inner` while the constructor for `compat::Interpreter` is called
https://github.com/compiler-research/CppInterOp/blob/c3384fb9d784a3f1606b6fc20c283c1e48d9b14b/lib/Interpreter/CppInterOpInterpreter.h#L153

which means `inner` can be `nullptr`

3) And then if we look at `CreateInterpreter` from the top, we're creating the `compat::Interpreter` here 

https://github.com/compiler-research/CppInterOp/blob/c3384fb9d784a3f1606b6fc20c283c1e48d9b14b/lib/Interpreter/CppInterOp.cpp#L2960

Which means `compat::Interpreter` can exist even when the underlying `clang::Interpreter` is not existing. So we can say that the Interpreter from the top is half broken

4) So if we try something like 
```
Cpp::CreateInterpreter(...)
Cpp::Parse(..)
```
We would hit the following finally
https://github.com/compiler-research/CppInterOp/blob/c3384fb9d784a3f1606b6fc20c283c1e48d9b14b/lib/Interpreter/CppInterOpInterpreter.h#L174-L176

And although the call is made to `compat::Interpreter`, we end up using the underlying `clang::Interpreter` **without checking if it is valid.**

## Type of change

Please tick all options which are relevant.

- [x] Bug fix
- [ ] New feature
- [ ] Requires documentation updates

## Testing

Please describe the test(s) that you added and ran to verify your changes.

## Checklist

- [x] I have read the contribution guide recently

anutosh491 commented:
Hence, we had a couple approaches to go ahead here

i) Introduce a isValid for clang::Interpreter and anytime we have a inner->func_call we use that (sounds a bit cumbersome checking for validity everytime)

ii) What Vipul and I liked was propagating the error all the way to that top, so that a half broken interpreter can't be put to use and a utility function like parse/declare would only work if CreateInterpreter doesn't return null along with an error message saying Interpreter creation failed.

anutosh491 commented on lib/Interpreter/Compatibility.h:
```diff
@@ -294,9 +294,11 @@ createClangInterpreter(std::vector<const char*>& args) {
+    if (auto Err = (*innerOrErr)->LoadDynamicLibrary("libcudart.so")) {
       llvm::logAllUnhandledErrors(std::move(Err), llvm::errs(),
                                   "Failed load libcudart.so runtime:");
+      return nullptr;
```
This was some inconsistency that I saw along the way. We return a nullptr whenever clang-repl uses ExitOnErr except this case (https://github.com/llvm/llvm-project/blob/main/clang/tools/clang-repl/ClangRepl.cpp#L212)

So clang-repl would just exit out ... through something like
```
(clang-repl-env-19) anutosh491@Anutoshs-MacBook-Air bin % ./clang-repl --cuda
clang-repl: dlopen(libcudart.so, 0x0009): tried: 'libcudart.so' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibcudart.so' (no such file), '/Users/anutosh491/micromamba/envs/clang-repl-env-19/lib/../lib/libcudart.so' (no such file), '/Users/anutosh491/micromamba/envs/clang-repl-env-19/lib/libcudart.so' (no such file), '/Users/anutosh491/micromamba/envs/clang-repl-env-19/bin/../lib/libcudart.so' (no such file), '/Users/anutosh491/micromamba/envs/clang-repl-env-19/bin/../lib/libcudart.so' (no such file), '/usr/lib/libcudart.so' (no such file, not in dyld cache), 'libcudart.so' (no such file), '/usr/local/lib/libcudart.so' (no such file), '/usr/lib/libcudart.so' (no such file, not in dyld cache)
```

But in our case we are continuing to use a half broken interpreter without the above working and we run declare on top of this 

https://github.com/compiler-research/CppInterOp/blob/c3384fb9d784a3f1606b6fc20c283c1e48d9b14b/unittests/CppInterOp/CUDATest.cpp#L17-L18

which is bound to not work as can be seen [here](https://github.com/compiler-research/CppInterOp/actions/runs/14242331470/job/39914920362#step:10:1001)

vgvassilev reviewed (approved):
Can we figure out how to write tests?

anutosh491 commented:
> Can we figure out how to write tests?

Okay I think the codecov says that 3 patches are not being tested. All of them are directly based out of the changes in `CXCppInterOp.cpp`

I tried adding tests but was quick to realize that I am not really doing just justice as I might not be the best one to goto for adding tests for the C API (hence I reverted them and squashed the commits 😅)

cc @Gnimuc 

Would be great if you could help me out with the tests here. The PR description (https://github.com/compiler-research/CppInterOp/pull/553#issue-2971536179) should be good to get you started on the changed made in the C-API (the changes are not too concerning, simple ones !)

1) Maybe for the creation of the Intepreter we could start with something simple like 
```
#ifndef CPPINTEROP_USE_CLING
TEST(InterpreterTest, CreateInterpreterCAPI) {
 const char* argv[] = {"-std=c++17"};
 CXInterpreter CXI = clang_createInterpreter(argv, 1);
 EXPECT_TRUE(CXI);


 clang_Interpreter_dispose(CXI);
}
#endif
```
2) For testing the failure while creating the interpreter (the case where we return `nullptr` from `clang_createInterpreter`), not sure what we can try. Maybe creating an non existent target or so I guess 
```
#ifndef CPPINTEROP_USE_CLING
TEST(InterpreterTest, CreateInterpreterCAPIFailure) {
  // This should return nullptr due to bad target triple
  const char* argv[] = {"-target", "nonexistent-unknown-target"};
  CXInterpreter CXI = clang_createInterpreter(argv, 2);
  EXPECT_EQ(CXI, nullptr);
}
#endif
```

anutosh491 commented:
> I'll give it a look this weekend.

Thanks a lot for the help. This change ends up being critical now that we'll soon be moving to llvm 20 !

Gnimuc commented:
The C API `clang_createInterpreter` aims to provide a way to customize compiler arguments that are used to create the compiler instance, so it doesn't adopt the following setups provided in the C++ API. 

https://github.com/compiler-research/CppInterOp/blob/98a57b4d31e487fb64bce223431d5249a369ff86/lib/Interpreter/CppInterOp.cpp#L2924-L2958

It's a bit tricky to set up a successful test environment because we need to know the path of system headers for each test platform. (Although most of the arguments are excluded in the C API,  [`-include new`](https://github.com/compiler-research/CppInterOp/blob/98a57b4d31e487fb64bce223431d5249a369ff86/lib/Interpreter/CppInterOpInterpreter.h#L151-L152) is mandatory, which needs the correct path of system headers)

Gnimuc commented:
Given `auto* I = Cpp::CreateInterpreter();`, if we can find a way to retrieve the arguments used to create the interpreter, then  those arguments can be used to test the C API.

anutosh491 commented:
Thanks for the reply @Gnimuc 

> BTW, I think the changes in the PR look good.

Also thanks for going through the changes ! Much appreciated :)

Hey @vgvassilev , do you think it's okay if we move ahead with this ?

As said above there's some discussion that needs to be done on testing the C API. And although some simple dummy tests can possibly be added as I wrote above, those won't probably be meaningful enough. So yeah the changes introduce **4 lines** of untested code in the C API but we can probably get to it in the future.

The main reason for this Pr is to help @mcbarton 's PR moving to llvm 20 (as his Pr should benefit with these changes after a rebase)

vgvassilev commented:
> The C API `clang_createInterpreter` aims to provide a way to customize compiler arguments that are used to create the interpreter, so it doesn't adopt the following setups provided in the C++ API.
> 
> https://github.com/compiler-research/CppInterOp/blob/98a57b4d31e487fb64bce223431d5249a369ff86/lib/Interpreter/CppInterOp.cpp#L2924-L2958
> 
> It's a bit tricky to set up a successful test environment because we need to know the path of system headers for each test platform. (Although most of the arguments are excluded in the C API, [`-include new`](https://github.com/compiler-research/CppInterOp/blob/98a57b4d31e487fb64bce223431d5249a369ff86/lib/Interpreter/CppInterOpInterpreter.h#L151-L152) is mandatory, which needs the correct path of system headers)

Can we make the new header optional?

anutosh491 commented:
> Can we make the new header optional?

I think this should be possible isn't it !

The include new header would be required for tests like these 
https://github.com/compiler-research/CppInterOp/blob/98a57b4d31e487fb64bce223431d5249a369ff86/unittests/CppInterOp/FunctionReflectionTest.cpp#L1361-L1375

Now on line `1375` we have the `GetAllTopLevelDecls` which ends up calling our `CreateInterpreter`

https://github.com/compiler-research/CppInterOp/blob/98a57b4d31e487fb64bce223431d5249a369ff86/unittests/CppInterOp/Utils.cpp#L21-L23

So technically if we can improve `GetAllTopLevelDecls` to take care of custom args-> we can pass the same args to `CreateInterpreter` from there and whichever test needs whatever header include can use it. That way it is not mandatory to get that include for tests who wouldn't want to use it.

But yeah that would mean changing the function definition of the utility function `GetAllTopLevelDecls` overall :\ 
So probably should be done in subsequent PR !!!

mcbarton commented:
> Ahh the 3 failing checks are due to this
> 
> [compiler-research/xeus-cpp#286](https://github.com/compiler-research/xeus-cpp/pull/286)
> 
> Making the same change here too ! I think we should be good after that !

I think this change should be made as a separate PR, and this PR rebased on it, given this pinning change has nothing to do with the goal of this PR. I will approve and merge this pinning PR, as soon as I see its been put in.

> > Can we make the new header optional?
> 
> I think this should be possible isn't it !
> 
> The include new header would be required for tests like these
> 
> https://github.com/compiler-research/CppInterOp/blob/98a57b4d31e487fb64bce223431d5249a369ff86/unittests/CppInterOp/FunctionReflectionTest.cpp#L1361-L1375
> 
> Now on line `1375` we have the `GetAllTopLevelDecls` which ends up calling our `CreateInterpreter`
> 
> https://github.com/compiler-research/CppInterOp/blob/98a57b4d31e487fb64bce223431d5249a369ff86/unittests/CppInterOp/Utils.cpp#L21-L23
> 
> So technically if we can improve `GetAllTopLevelDecls` to take care of custom args-> we can pass the same args to `CreateInterpreter` from there and whichever test needs whatever header include can use it. That way it is not mandatory to get that include for tests who wouldn't want to use it.
> 
> But yeah that would mean changing the function definition of the utility function `GetAllTopLevelDecls` overall :\ So probably should be done in subsequent PR !!! (which I can try tomorrow)

If making this header option would allow this PR to be tested, then should probably be done first, and then this PR should be rebased this on it. Otherwise we risk tests never been made to cover these lines. I know its just a few lines, but its because due to not insisting all lines be covered by tests, we have ended up in a situation where CppInterOps coverage has dropped by approximately 13% in the last 12 months.

anutosh491 commented:
Hi,

> I think this change should be made as a separate PR

I've taken care of this here https://github.com/compiler-research/CppInterOp/pull/559

> If making this header option would allow this PR to be tested, then should probably be done first, and then this PR should be rebased this on it. Otherwise we risk tests never been made to cover these lines. I know its just a few lines, but its because due to not insisting all lines be covered by tests, we have ended up in a situation where CppInterOps coverage has dropped by approximately 13% in the last 12 months (I made an issue about this here https://github.com/compiler-research/CppInterOp/issues/541).

Hmmm, I get it but shouldn't higher  priority be moving to LLVM 20 now (it was out on FEB 2 and its high time we move) . The purpose of this PR was to help with that. 

Also I am not a huge fan of nobody officially taking responsibility for covering the C API. It shouldn't have been added in the first place if that was the case !

Most of my focus since the past 2-3 weeks is around getting a good build (native/emscripten) for cppinterop around llvm 20 and technically this PR is also tested well on our C++ API. But for the C-API
i) I am not educated enough on the use case.
ii) I would really fall short on time to handle development/coverage there.

But this shouldn't slow any development on C++ API side is what I'm trying to convey.

I'm just going by Gnimuc 's word who added the C-API (he says "BTW, I think the changes in the PR look good." https://github.com/compiler-research/CppInterOp/pull/553#issuecomment-2799855918) so I am just hoping we would be in a good enough state to test it in the near future. But if something as simple as a "create interpreter" function is not tested in the C API do we really stop other work due to that ? I don't know.

Gnimuc commented:
> #ifndef CPPINTEROP_USE_CLING
> TEST(InterpreterTest, CreateInterpreterCAPIFailure) {
>   // This should return nullptr due to bad target triple
>   const char* argv[] = {"-target", "nonexistent-unknown-target"};
>   CXInterpreter CXI = clang_createInterpreter(argv, 2);
>   EXPECT_EQ(CXI, nullptr);
> }
> #endif

@anutosh491 Maybe it's a good idea to add this failure test for now. It will improve the coverage a bit.

Gnimuc commented:
> > The C API `clang_createInterpreter` aims to provide a way to customize compiler arguments that are used to create the interpreter, so it doesn't adopt the following setups provided in the C++ API.
> > https://github.com/compiler-research/CppInterOp/blob/98a57b4d31e487fb64bce223431d5249a369ff86/lib/Interpreter/CppInterOp.cpp#L2924-L2958
> > 
> > It's a bit tricky to set up a successful test environment because we need to know the path of system headers for each test platform. (Although most of the arguments are excluded in the C API, [`-include new`](https://github.com/compiler-research/CppInterOp/blob/98a57b4d31e487fb64bce223431d5249a369ff86/lib/Interpreter/CppInterOpInterpreter.h#L151-L152) is mandatory, which needs the correct path of system headers)
> 
> Can we make the new header optional?

I'm wondering what's the original purpose of adding this header. Is it related to the hard-coded code in `clang::Interpreter::create`?  An operator new trampoline function is declared there.

anutosh491 commented:
Hey @Gnimuc, we just dropped the header through #https://github.com/compiler-research/CppInterOp/pull/561

Does this put you in a better position to add some basic tests for `clang_createInterpreter` or other related functions ?

anutosh491 commented on unittests/CppInterOp/InterpreterTest.cpp:
```diff
@@ -166,6 +166,25 @@
+  CXInterpreter CXI = clang_createInterpreter(argv, 1);
+  EXPECT_TRUE(CXI);
+
+  auto CLI = clang_Interpreter_getClangInterpreter(CXI);
```
I see "auto CLI" being put to use through while getting access to the Interpreter. So hope that's fine.

anutosh491 commented on unittests/CppInterOp/InterpreterTest.cpp:
```diff
@@ -167,6 +167,22 @@ TEST(InterpreterTest, CreateInterpreter) {
+}
+
+TEST(InterpreterTest, CreateInterpreterCAPIFailure) {
+  const char* argv[] = {"-fsyntax-only", "-Xclang", "-invalid-plugin"};
```
https://github.com/compiler-research/CppInterOp/pull/553/files#r2043868646

mcbarton commented on unittests/CppInterOp/InterpreterTest.cpp:
```diff
@@ -167,6 +167,22 @@ TEST(InterpreterTest, CreateInterpreter) {
 #endif
 }
 
+#ifndef CPPINTEROP_USE_CLING
```
Is there a reason why we are not doing this test for Cling?

anutosh491 replied:
No clue,

I see no C-API tests done through cling and that's what I am trying to replicate. Maybe @Gnimuc can educate us here.

Gnimuc replied:
The original purpose of adding the C API is to support multiple instances. REPL and Cling are two backends for creating compiler instances. REPL is based on `clang::Interpreter` from the upstream llvm-project. 

I think there is no plan to upstream Cling to the llvm-project, and I didn't check Cling's compatibility when I added those C-API tests.

vgvassilev replied:
You can build, CppInterOp by providing cling and it will embed cling in it. If that’s not too much of work maybe we can support that, shouldn’t hurt.

anutosh491 replied:
Can we have a dedicated PR for that ? 

I'll say the purpose of this PR is complete now.

>  If that’s not too much of work maybe we can support that, shouldn’t hurt.

@Gnimuc would you like to give this an attempt ?

kr-2003 commented:
> Hey @Gnimuc @vgvassilev
> 
> I have added some simple tests and ensured codecov doesn't complain about changes atleast with regard to this PR.
> 
> All tests pass no clue why the windows CI fail (don't look like a direct outcome of my tests)
> 
> ![image](https://private-user-images.githubusercontent.com/87052487/433735494-3f563b21-aae6-41d2-b320-cb0c77c1edcc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ3MDU5MzEsIm5iZiI6MTc0NDcwNTYzMSwicGF0aCI6Ii84NzA1MjQ4Ny80MzM3MzU0OTQtM2Y1NjNiMjEtYWFlNi00MWQyLWIzMjAtY2IwYzc3YzFlZGNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE1VDA4MjcxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdjZDVlYjU5NGM4ZDc5NWRkOTgwMTNjMTk3OGJiNTE0NjI4YjY5YWNmYTkzNjE0MGE3OWJiM2Q4Y2E0MjQ5ZDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.3PePusgqNIJdLO2A7ERtWVbDRBHfxfX9T5yn7PHzuZ4)
> 
> Shall try looking into it for a bit (or shall skip If I am helpless cause I don't have a windows machine :\ )

Tests which write something on screen fails on Windows. Here, in this comment @mcbarton explained this:
https://github.com/compiler-research/CppInterOp/pull/513#issuecomment-2735698456

mcbarton commented:
> > Hey @Gnimuc @vgvassilev
> > I have added some simple tests and ensured codecov doesn't complain about changes atleast with regard to this PR.
> > All tests pass no clue why the windows CI fail (don't look like a direct outcome of my tests)
> > ![image](https://private-user-images.githubusercontent.com/87052487/433735494-3f563b21-aae6-41d2-b320-cb0c77c1edcc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ3MDU5MzEsIm5iZiI6MTc0NDcwNTYzMSwicGF0aCI6Ii84NzA1MjQ4Ny80MzM3MzU0OTQtM2Y1NjNiMjEtYWFlNi00MWQyLWIzMjAtY2IwYzc3YzFlZGNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDE1VDA4MjcxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdjZDVlYjU5NGM4ZDc5NWRkOTgwMTNjMTk3OGJiNTE0NjI4YjY5YWNmYTkzNjE0MGE3OWJiM2Q4Y2E0MjQ5ZDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.3PePusgqNIJdLO2A7ERtWVbDRBHfxfX9T5yn7PHzuZ4)
> > Shall try looking into it for a bit (or shall skip If I am helpless cause I don't have a windows machine :\ )
> 
> Tests which write something on screen fails on Windows. Here, in this comment @mcbarton explained this: [#513 (comment)](https://github.com/compiler-research/CppInterOp/pull/513#issuecomment-2735698456)

To refine this comment from before. I think anything which writes an error message in Windows, will be picked up as an error in the overall running of the tests. Tests which seem to write a warning message are ok. I believe it is ok for you skip this test on Windows for this reason.

kr-2003 commented:
> To refine this comment from before. I think anything which makes the interpreter write an error message in Windows, will be picked up as an error in the overall running of the tests. Tests which seem to write a warning message are ok. I believe it is ok for you skip this test on Windows for this reason.

Alternatively, we can suppress the error message and retain this test on windows too.
```

TEST(InterpreterTest, CreateInterpreterCAPIFailure) {
#ifdef CPPINTEROP_USE_CLING
  GTEST_SKIP() << "C API is not available in this build";
#endif
  const char* argv[] = {"-fsyntax-only", "-Xclang", "-invalid-plugin"};
  testing::internal::CaptureStderr(); // Suppress error messages
  auto *CXI = clang_createInterpreter(argv, 3);
  std::string capturedStderr = testing::internal::GetCapturedStderr();
  EXPECT_EQ(CXI, nullptr);
}
```

mcbarton commented:
> > To refine this comment from before. I think anything which makes the interpreter write an error message in Windows, will be picked up as an error in the overall running of the tests. Tests which seem to write a warning message are ok. I believe it is ok for you skip this test on Windows for this reason.
> 
> Alternatively, we can suppress the error message and retain this test on windows too. It should pass now.
> 
> ```
> 
> TEST(InterpreterTest, CreateInterpreterCAPIFailure) {
> #ifdef CPPINTEROP_USE_CLING
>   GTEST_SKIP() << "C API is not available in this build";
> #endif
>   const char* argv[] = {"-fsyntax-only", "-Xclang", "-invalid-plugin"};
>   testing::internal::CaptureStderr(); // Suppress error messages
>   auto *CXI = clang_createInterpreter(argv, 3);
>   std::string capturedStderr = testing::internal::GetCapturedStderr();
>   EXPECT_EQ(CXI, nullptr);
> }
> ```

Not entirely sure testing::internal::CaptureStderr() will work to resolve this issue. Its probably better to skip for now on Windows. We do however want to use GTEST_SKIP(), instead of the way its done currently in the PR.

anutosh491 commented:
> Its probably better to skip for now on Windows

Yeah I think we can just skip it for now.

anutosh491 commented:
I think this is ready from my side now that all the checks pass.

vgvassilev commented on unittests/CppInterOp/CUDATest.cpp:
```diff
@@ -30,7 +30,7 @@ static bool HasCudaRuntime() {
       return false;
 
-    Cpp::CreateInterpreter({}, {"--cuda"});
+    if(!Cpp::CreateInterpreter({}, {"--cuda"})) return false;
```
```suggestion
    if (!Cpp::CreateInterpreter({}, {"--cuda"})) 
      return false;
```

why clang-format was happy with this? Please fix the other occurrences too.

anutosh491 replied:
Oops completely missed this (not sure I saw a complaint regarding this !

Shall address it, thanks

vgvassilev reviewed (approved):
Please also address the missing includes suggested by clang-tidy before merging.

anutosh491 commented:
Addressed all reviews. I think this should be ready now. Thanks everyone for the reviews. Squashing and Merging!
